sync15-adapter/src/token.rs

```diff
@@ -130,4 +148,84 @@ impl TokenserverClient {
 
         Ok(Authorization(format!("Hawk {}", header)))
     }
+
+    pub fn api_endpoint(&self) -> String {
+        self.api_endpoint.clone()
+    }
+}
+
+
+/// The TokenserverClient - long lived and fetches tokens on demand (eg, when
+/// first needed, or when an existing one expires.)
+#[derive(Debug)]
+pub struct TokenserverClient {
+    // The stuff needed to fetch a token.
+    base_url: String,
+    access_token: String,
+    key_id: String,
+    // The token (and derived info) that we fetch and hand back to consumers.
+    // While this seems odd, we use an Rc to try and ensure consumers can't
+    // hold on to it - we'll panic if we try and update it.
+    // (Also because we can't clone an TokenserverContext as hawk doesn't derive clone,
+    // but even if we could, we'd run the risk of a consumer using a stale
+    // clone. No idea if this actually makes sense though.)
+    current_context: RefCell<Option<Rc<TokenserverContext>>>,
```

Comment:
Is there ever a case where `current_context.is_some() && current_backoff.is_some()`? If not, I'm wondering if we could formalize these states in an enum, something like:

```rust
enum TokenserverContextState {
  NotFetched,
  Current(TokenserverContext),
  BackoffUntil(SystemTime)
}
```

And keep a `RefCell` to that.

Reply:
Not a fan of this. In general using `Rc` and `RefCell` are a sign you aren't really writing idiomatic rust. In this case, I don't think this does what you want either (specifically, it won't actually allow you to prevent stale use).

We can't clone a Credentials but we actually can recreate them from the token, which we already have around. Storing the credentials is basically just an optimization that I wouldn't have done if I realized the Credential object had bad ergonomics like that. 

As I mentioned, this won't prevent using a stale context. The type you probably want is `Rc<RefCell<Option<TokenServerContext>>>`, which *would* let us update the callers context, however it would panic if we tried to do it while they were borrowing it. However this seems very likely to cause problems, for example, we have no guarantee they aren't borrowing the context when we decide to update it for them, which could cause a panic.

I'm pretty sure the right thing here probably something like what kit is proposing, and then hold onto a `RefCell<TokenServerContextState>` so that we can have interior mutability. Then, I don't think get_context should be a public method, or if it is it should return `Result<RefMut<TokenServerContext>>` (but I think it shouldn't be one).

Reply:
Thanks for the great feedback - most of the comments made are now stale (so I haven't replied to them all) as the new version of this uses an enum as Kit proposes, while still sticking with Thom's (very helpful) refactor to use a callback. This has actually made it quite nice IMO.

I'm still struggling a little here with a state that represents there was an error fetching the token (and specifically, how to get this error back to the caller), but I figured I'd get a new version up now for general feedback and to gather some ideas for that error state - they are fairly easy to spot with "todo" etc comments.

Another thing I gave up on was testing using traits, so suggestions for how to simply test are also invited (although it also seems reasonable we just open an issue for tests and accept a PR without them given none exist today)